packages/expo-screen-orientation/ios/EXScreenOrientation/EXScreenOrientationUtilities.m

```diff
@@ -0,0 +1,170 @@
+// Copyright 2019-present 650 Industries. All rights reserved.
+
+#import <Foundation/Foundation.h>
+#import <EXScreenOrientation/EXScreenOrientationUtilities.h>
+
+#import <sys/utsname.h>
+
+@implementation EXScreenOrientationUtilities
+
+# pragma mark - helpers
+
++ (BOOL)doesSupportOrientationMask:(UIInterfaceOrientationMask)orientationMask
+{
+  if ((UIInterfaceOrientationMaskPortraitUpsideDown & orientationMask) // UIInterfaceOrientationMaskPortraitUpsideDown is part of orientationMask
+      && [self doesDeviceHaveNotch])
+  {
+    // device does not support UIInterfaceOrientationMaskPortraitUpsideDown and it was requested via orientationMask
+    return false;
+  }
+  
+  return true;
+}
+
++ (BOOL)doesDeviceHaveNotch {
+  if (@available(iOS 11.0, *)) {
+    __block BOOL result = false;
+    dispatch_semaphore_t sem = dispatch_semaphore_create(0);
+    dispatch_async(dispatch_get_main_queue(), ^{
```

Comment:
1. Let's make it remember the value, so we block the execution only once.
2. Does `top` always resolve to `portrait-top`, or does it return `landscapeLeft-top` when running the function in landscape mode?
3. Why `> 20.0`? Would `> 0.0` work too? 
4. Do you think this is more reliable than the old way, based on `systemInfo.machine`?
5. What would happen if the status bar was hidden?
6. I have also seen detecting notch by checking… bottom safe area inset. Not saying we should migrate to it, just noting that there is such way.

Reply:
You were right, this doesn't always resolve to `portrait-top`. I've changed it. Now, it's detecting notch base on bottom safe area inset. 

I think it's more maintainable then `systemInfo.machine`.